rfd/xxxx-headless-authentication.md

```diff
@@ -0,0 +1,386 @@
+
+Example:
+
+```
+Incoming headless request for command "tsh ssh server01", to connect to node <server01_uuid>. Upon approval, the request will be awarded the following permissions:
+
+allow:
+  options:
+    max_session_ttl: 8h0m0s
+    forward_agent: true
+    port_forwarding: true
+    require_session_mfa: true
+  logins: ["ubuntu-user"]
+  resource_ids: ["<server01_uuid>"]
+  rules:
+    resources: ["nodes"]
+    verbs: ["read", "list"]
+```
+
+**Important note**: The server cannot guarantee that the client will actually use the resulting certificates for the `tsh` command advertised by the client. This means that a modified client could claim to need certificates for `tsh ls`, but then actually use the certificates for `tsh ssh`. In order to prevent potential permission workarounds, we will need to carefully consider how each supported command's awarded certificates may interact with other Teleport clients. For this reason, we may add more commands as they are requested, but each command may have significant security implications to consider. Each PR adding one or more new commands should include a security description and get a security review from the Teleport security team. This includes the commands listed above, as the most limited necessary permissions necessary to perform each command is not yet clear.
+
+#### Caveat
+
+Due to the implementation complexity of limiting certificates permissions by requested command, it will not be included in the feature's preview release.
+
+Additionally, some details, like how the server will encode the permissions into the certificate, are currently missing. This will be addressed in a follow up PR before the official release of this feature, accompanied by an RFD update.
+
+The feature will still be limited to `tsh ls`, `tsh ssh`, and `tsh scp` so that we don't need to remove functionality later.
+
+#### Conclusion
+
+With the design principles above, the only possible attack would be for the attacker to issue the headless request themselves and trick the user into verifying the request with MFA, with a direct phishing attack. This phishing attack should also be mitigated by notifying the user of the command requested and the permissions that will be awarded upon approval.
+
+### Headless authentication overview
+
+The Headless authentication flow will be modeled after our existing SSO authentication flows, where the user issues an `AuthRequest`, directs the client to approve the request via SSO, and then completes authentication with the approved auth request. Unlike with SSO authentication, we have control over the headless authentication flow through our Proxy/Auth APIs.
+
+#### Diagram
+
+The headless authentication flow is shown below:
+
+```mermaid
+sequenceDiagram
+    participant Local Machine
+    participant Remote Machine
+    participant Teleport Proxy
+    participant Teleport Auth
+    
+    Note over Remote Machine: tsh --headless ssh user@node01
```

Comment:
Did we consider requiring `--user` for headless mode? It's important for the user and request to be tightly coupled and personal env variables in remote machines strike me wrong.

Reply:
The user would either need to provide `--user` or set `$TELEPORT_USER`. The latter is included in the UX section, but if we have a security concern greater than the UX concern here, we can remove it.